### PR TITLE
SlidingWindow Cleanup

### PR DESCRIPTION
**Pull request by jklontz**
@sklum Here's my first pass through the code. This just does the obvious stuff like hoisting invariants outside of loops, removing redundant variables, improving const correctness, etc. This reduced the runtime on my test case by 30% (from 83 to 58 seconds). Comments inline. @JordanCheney FYI.

### CODE REVIEW

File: openbr/plugins/imgproc/slidingwindow.cpp [hunk 1]
```diff
@@ -91,50 +91,44 @@ class SlidingWindowTransform : public MetaTransform
...
-                Size windowSize(cvRound(originalWindowSize.width*factor), cvRound(originalWindowSize.height*factor) );
-                Size scaledImageSize(cvRound(m.cols/factor ), cvRound(m.rows/factor));
-                Size processingRectSize(scaledImageSize.width - originalWindowSize.width, scaledImageSize.height - originalWindowSize.height);
+                const Size windowSize(cvRound(originalWindowSize.width*factor), cvRound(originalWindowSize.height*factor));
+                const Size scaledImageSize(cvRound(imageSize.width/factor), cvRound(imageSize.height/factor));
+                const Size processingRectSize(scaledImageSize.width - originalWindowSize.width, scaledImageSize.height - originalWindowSize.height);
 
                 if (processingRectSize.width <= 0 || processingRectSize.height <= 0)
                     break;
-                if (windowSize.width > maxObjectSize.width || windowSize.height > maxObjectSize.height)
```
Comment (jklontz) [inline openbr/plugins/imgproc/slidingwindow.cpp]: It seems that this condition is impossible to reach as the one above will always be true first?

File: openbr/plugins/imgproc/slidingwindow.cpp [hunk 2]
```diff
@@ -91,50 +91,44 @@ class SlidingWindowTransform : public MetaTransform
...
-                    scaleBuffer.append(scaledImage);
+                    resize(m, scaledImage, scaledImageSize, 0, 0, CV_INTER_LINEAR);
+                    rep.append(scaledImage);
                 }
+                rep = classifier->preprocess(rep);
 
-                Template rep(t.file);
-                rep = classifier->preprocess(scaleBuffer);
+                // Pre-allocate the window to avoid constructing this every iteration
+                Template window(t.file);
```
Comment (jklontz) [inline openbr/plugins/imgproc/slidingwindow.cpp]: While I didn't test this optimization in isolation, this is probably the primary source of speedup.

File: openbr/plugins/imgproc/slidingwindow.cpp [hunk 3]
```diff
@@ -155,12 +149,12 @@ class SlidingWindowTransform : public MetaTransform
...
 
             if (!enrollAll && rects.empty()) {
-                rects.append(Rect(0, 0, m.cols, m.rows));
+                rects.append(Rect(0, 0, imageSize.width, imageSize.height));
                 confidences.append(-std::numeric_limits<float>::max());
             }
 
             for (int j=0; j<rects.size(); j++) {
-                Template u(t.file, m);
+                Template u = t;
```
Comment (jklontz) [inline openbr/plugins/imgproc/slidingwindow.cpp]: I assume you meant to propagate all the matrices?

**Comment by sklum** [comment]
lgtm

File: openbr/plugins/imgproc/slidingwindow.cpp (hunk 3 above)
Comment (sklum) [inline openbr/plugins/imgproc/slidingwindow.cpp]: Yup.

**Pull request merged**